File: packages/sdk/src/observability/otel.ts
/**
 * Optional OpenTelemetry integration.
 *
 * Uses dynamic import so @opentelemetry/api is a zero-cost optional peer dep.
 * If not installed, all functions return no-ops.
 */

export interface VetoTracer {
  startSpan(name: string): VetoSpan;
}

export interface VetoSpan {
  setAttribute(key: string, value: string | number | boolean): void;
  setStatus(status: { code: number; message?: string }): void;
  end(): void;
}

// Canonical OTEL status codes (avoid importing from @opentelemetry/api directly)
export const SpanStatusCode = { OK: 1, ERROR: 2 } as const;

const noopSpan: VetoSpan = {
  setAttribute() {},
  setStatus() {},
  end() {},
};

const noopTracer: VetoTracer = {
  startSpan() { return noopSpan; },
};

/**
 * Attempt to load @opentelemetry/api. Returns a VetoTracer wrapping the real
 * tracer, or a no-op tracer if the package is not installed.
 *
 * Call once during Veto.init() and store the result.
 */
export async function tryLoadOtel(serviceName = 'veto-sdk'): Promise<VetoTracer> {
  try {
    // Dynamic import — works in Node.js 20+ ESM. Throws ERR_MODULE_NOT_FOUND if not installed.
    const api = await import('@opentelemetry/api');
    const tracer = api.trace.getTracer(serviceName);
    return {
      startSpan(name: string): VetoSpan {
        const span = tracer.startSpan(name);
        return {
          setAttribute(key, value) { span.setAttribute(key, value); },
          setStatus(status) { span.setStatus(status); },
          end() { span.end(); },
        };
      },
    };
  } catch {
     
    console.debug('veto: @opentelemetry/api not available, tracing disabled');
    return noopTracer;
  }
}

File: packages/sdk-python/veto/observability/otel.py
"""
Optional OpenTelemetry integration.

Uses dynamic import so opentelemetry-api is a zero-cost optional dep.
If not installed, all functions return no-ops.
"""

from typing import Protocol, Union


class VetoSpan(Protocol):
    def set_attribute(self, key: str, value: Union[str, int, bool]) -> None: ...
    def set_status(self, status_code: int, message: str = "") -> None: ...
    def end(self) -> None: ...


class VetoTracer(Protocol):
    def start_span(self, name: str) -> VetoSpan: ...


SPAN_STATUS_OK = 1
SPAN_STATUS_ERROR = 2


class _NoopSpan:
    def set_attribute(self, key: str, value: Union[str, int, bool]) -> None:
        pass

    def set_status(self, status_code: int, message: str = "") -> None:
        pass

    def end(self) -> None:
        pass


class _NoopTracer:
    def start_span(self, name: str) -> VetoSpan:
        return _NoopSpan()


def try_load_otel(service_name: str = "veto-sdk") -> VetoTracer:
    """Load opentelemetry.api if installed, else return noop tracer."""
    try:
        from opentelemetry import trace

        tracer = trace.get_tracer(service_name)

        class _OtelSpan:
            def __init__(self, span: trace.Span) -> None:
                self._span = span

            def set_attribute(self, key: str, value: Union[str, int, bool]) -> None:
                self._span.set_attribute(key, value)

            def set_status(self, status_code: int, message: str = "") -> None:
                from opentelemetry.trace import StatusCode, Status

                sc = StatusCode.OK if status_code == SPAN_STATUS_OK else StatusCode.ERROR
                self._span.set_status(Status(sc, message))

            def end(self) -> None:
                self._span.end()

        class _OtelTracer:
            def start_span(self, name: str) -> VetoSpan:
                return _OtelSpan(tracer.start_span(name))

        return _OtelTracer()
    except ImportError:
        return _NoopTracer()
